# Lab Exercises

## Introduction to Python, Part 3

### Questions/Office Hours

If you have questions about this week's lab **ANY** TA for this class can help you with any problems in office hours. That being said, if you need help over email it is best to email **YOUR** TA.

You can also always ask the instructor for help with skills labs (even though they are not covered in lecture). You can email the instructor of come to his office hours.

----

## Exercises

The exercises this week deal with some features that we haven't otherwise covered. There are links to specific parts of the Python for Everyone book in each question, but here's a more general reference:

* [Chapter 7: Files](https://www.py4e.com/html3/07-files)
* [Chapter 8: Lists](https://www.py4e.com/html3/08-lists)
* [Chapter 9: Dictionaries](https://www.py4e.com/html3/09-dictionaries)

### A Note on Boilerplate

One of the major benefits of tools like Jupyter are that it makes it easy to strip away the "boilerplate" that comes with programming languages; some of these exercises relies on a little bit more of that than in weeks past. If you get confused or something stops working, email your TA or come to ofice hours.

----

#### **Exercise 1**

###### sections to reference: [opening files](https://www.py4e.com/html3/07-files)

This one definitely falls in the category of "pointless exercises". You're going to make a program that prompts for a filename. If the file exists, quit the program! Don't print anything out, don't do anything with the file, don't do anything. Just quit. If the file *doesn't* exist, complain, and prompt the user for a filename again. Do this until the user gives you a valid filename or pours water on their computer in frustration. (Just kidding. Don't pour water on your computer.)

You're going to want to put the filename prompt inside a loop, so you can ask the user for a filename as many times as you want. But what kind of loop? And how will you get out of the loop once you've got a valid file?

![thinking face](https://media.giphy.com/media/3o7buirYcmV5nSwIRW/giphy.gif)

In [4]:
import os.path
from os import path

notvalid = True
while(notvalid):
    fn = input(prompt="Enter in a file name: ")
    if path.isfile(fn):
        notvalid = False

Enter in a file name: C:\Users\tatty\Downloads\c2w2.pdf


#### **Exercise 2**

###### sections to reference: [opening files](https://www.py4e.com/html3/07-files)

Write a program that asks the user for a file and simply prints the entire contents of the file out. If the user enters a file that doesn't exist, give the user an error and prompt for a filename again. 

**Hint**: You can do the printing in *one line of code*.

In [6]:

notvalid = True
while(notvalid):
    fn = input(prompt="Enter in a file path to a file to have it file contents printed out: ")
    if path.isfile(fn):
        notvalid = False
#path is valid now we want to print its contents out
#open the file and store it contents
contents = open(fn)
#get all the lines of the file read in
allLines = contents.readlines()
#iterate over all the lines and print them out
for line in allLines:
    print(line)

Enter in a file path to a file to have it file contents printed out: C:\Users\tatty\Downloads\test.txt.txt
test1

test2 test3

test4


#### **Exercise 3**

###### sections to reference: [parsing strings](https://www.py4e.com/html3/06-strings)

This exercise deals with a test file, `spam1.txt`. You can open that file in a new tab so you can get a sense for what it looks like and what you're going to be dealing with. 

You're going to be writing a program that deals with a list of spam confidence values. The file, spam1.txt, looks something like this:

There are 500 spam confidence values in the file `spam1.txt`. I want you to tell me two things about the values:

1. The **average** of all of the spam values in the file, and
1. **How many** emails have a spam value of over 95% (so, >0.95).

You'll need to read the file line by line and process each line so you can get the float.

After you're done with that program, make a Markdown cell below your code and **tell me**, in a few sentences, what the average spam confidence *means*. By way of explanation, the spam value is a value generated by an algorithm (individual email providers keep their algorithms closely guarded, but they're all pretty much based on *Bayesian math*, if you're interested.) That algorithm rates any given email from 0 to 1 on how confident they are that the email is spam. 1 means that the algorithm thinks it *definitely* is spam, and 0 means that the algorithm thinks there's no chance it's spam.

So, once you have your average value, **what does that mean** for any given email in the set? What are the chances it'll be spam or not?

In [18]:
spam1 = open("C:/Users/tatty/0010Github/Python-Part3/spam1.txt")
EO95 = 0
numOfValues = 0
totalOfValues = 0
#get all the lines of the file read in
allLines = spam1.readlines()
#iterate over all the lines and print them out
for line in allLines:
    splitNum = line.split(":")[1]
    splitNum = float(splitNum)
    if(splitNum > .95):
        EO95+=1
    totalOfValues += float(splitNum)
    numOfValues +=1
print("Analysis of spam1.txt:\n")
print("Emails with a spam value over 95%: ", EO95)
print("Percent of emails with a value over 95%: ", (EO95 / numOfValues) * 100,"%")
print("Number of values in the file: ", numOfValues)
print("Total of all the values in the file:", totalOfValues)
print("Avg spam confidence of spam1.txt: ", totalOfValues / numOfValues)


Analysis of spam1.txt:

Emails with a spam value over 95%:  68
Percent of emails with a value over 95%:  13.600000000000001 %
Number of values in the file:  500
Total of all the values in the file: 375.3995999999997
Avg spam confidence of spam1.txt:  0.7507991999999993


Analysis of spam1.txt:

Emails with a spam value over 95%:  68
Percent of emails with a value over 95%:  13.600000000000001 %
Number of values in the file:  500
Total of all the values in the file: 375.3995999999997
Avg spam confidence of spam1.txt:  0.7507991999999993
# What the values mean
With a avg spam confidence of spam1.txt data we see a value of 0.75079 or 0.7508 / 0.100. This means that over two thirds of the emails had a spam confidence value of atleast 0.75079. The spam confidence value is a rating of how confident the algorithm is that the email is spam. So over all in the spam1 we see a good amount of spam. IE you have a 3/4 chance of opening spam for every email that exists in this set.

#### **Exercise 4**

###### sections to reference: [searching through a file](https://www.py4e.com/html3/07-files)

This exercise is a continuation of the previous one. You've got a new text file to process, called `spam2.txt`. It has a bunch of spam confidence values, but it's also got some other information, too, like the sender, message subject, and the date and time that it's been sent. A sample record from the file looks like this:

There is a lot of information here, and this, if you can believe it, is a *stripped-down* version of what the email server records look like. 

`spam2.txt` contains a bunch of email records. You're going to write a program, similar to Exercise 3, wherein you grab the spam confidence from each email. Like before, you're going to **take the average** of these values. When you encounter an email with a spam confidence greater than 95%, though, you're going to **print out** the email address that sent the email, the email address that received it, the date and time that it was sent, and the spam confidence value. 

**Note**: Don't simply print out the entire lines of the file that have that information; get the specific items that you want and format it nicely. 

In [56]:
spam2 = open("C:/Users/tatty/0010Github/Python-Part3/spam2.txt")
EO95 = 1
numOfValues = 1
totalOfValues = 0
#get all the lines of the file read in
allLines = spam2.readlines()
#iterate over all the lines and print them out
for line in allLines:
    #print("Current Line: ", line)
    if "From" in line:
        From = line
        #print("From line: ",From)
    elif "To:" in line:
        to = line
        #print("To line: ", to)
    elif "Date:" in line:
        date = line
        #print("Date: ",date)
    elif "X-DSPAM-Confidence:" in line:
        splitNum = line.split(":")[1]
        #print("Confidence value: ", splitNum)
        splitNum = float(splitNum)
        totalOfValues += float(splitNum)
        numOfValues +=1
        if(splitNum > .95):
            EO95+=1
            print("----------------------------------------------------------------------------------------------")
            print("High Confidence value found:\n"+ From+"\n"+to,"\n"+date,"\nXSConfidence Val: ",splitNum,"\n")
            print("----------------------------------------------------------------------------------------------")
        
print("Analysis of spam2.txt:\n")
print("----------------------------------------------------------------------------------------------")
print("Emails with a spam value over 95%: ", EO95)
print("Percent of emails with a value over 95%: ", (EO95 / numOfValues) * 100,"%")
print("Number of values in the file: ", numOfValues)
print("Total of all the values in the file:", totalOfValues)
print("Avg spam confidence of spam2.txt: ", totalOfValues / numOfValues)
print("Percent Avg spam confidence of spam2.txt:", (totalOfValues / numOfValues) * 100,"%")
print("----------------------------------------------------------------------------------------------")

#NOTES:
#Analysis of spam2.txt:
#----------------------------------------------------------------------------------------------
#Emails with a spam value over 95%:  1003
#Percent of emails with a value over 95%:  55.784204671857616 %
#Number of values in the file:  1798
#Total of all the values in the file: 1606.7480999999987
#Avg spam confidence of spam2.txt:  0.8936307563959948
#Percent Avg spam confidence of spam2.txt: 89.36307563959947 %
#----------------------------------------------------------------------------------------------

----------------------------------------------------------------------------------------------
High Confidence value found:
From: cwen@iupui.edu

To: source@collab.sakaiproject.org
 
Date: Thu, 3 Jan 2008 16:33:03 -0500
 
XSConfidence Val:  0.9846 

----------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------
High Confidence value found:
From: cwen@iupui.edu

To: source@collab.sakaiproject.org
 
Date: Thu, 3 Jan 2008 16:22:15 -0500
 
XSConfidence Val:  0.9907 

----------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------
High Confidence value found:
From: zqian@umich.edu

To: source@collab.sakaiproject.org
 
Date: Thu, 3 Jan 2008 15:25:57 -0500
 
XSConfidence Val:  0.9895 

-----------------------------------------------------------

----------------------------------------------------------------------------------------------
High Confidence value found:
From: wagnermr@iupui.edu

To: source@collab.sakaiproject.org
 
Date: Mon, 17 Dec 2007 14:14:27 -0500
 
XSConfidence Val:  0.9894 

----------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------
High Confidence value found:
From: dlhaines@umich.edu

To: source@collab.sakaiproject.org
 
Date: Mon, 17 Dec 2007 13:37:15 -0500
 
XSConfidence Val:  0.9861 

----------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------
High Confidence value found:
From: cwen@iupui.edu

To: source@collab.sakaiproject.org
 
Date: Mon, 17 Dec 2007 13:30:31 -0500
 
XSConfidence Val:  0.9819 

-------------------------------------------------

XSConfidence Val:  0.9824 

----------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------
High Confidence value found:
From: chmaurer@iupui.edu

To: source@collab.sakaiproject.org
 
Date: Fri, 7 Dec 2007 15:31:08 -0500
 
XSConfidence Val:  0.9809 

----------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------
High Confidence value found:
From: chmaurer@iupui.edu

To: source@collab.sakaiproject.org
 
Date: Fri, 7 Dec 2007 15:39:14 -0500
 
XSConfidence Val:  0.9783 

----------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------
High Confidence value found:
From: zqian@umich.edu

To: source@collab.sakaiproject.o

XSConfidence Val:  0.9849 

----------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------
High Confidence value found:
From: chmaurer@iupui.edu

To: source@collab.sakaiproject.org
 
Date: Mon, 19 Nov 2007 08:59:54 -0500
 
XSConfidence Val:  0.9849 

----------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------
High Confidence value found:
From: aaronz@vt.edu

To: source@collab.sakaiproject.org
 
Date: Mon, 19 Nov 2007 02:36:41 -0500
 
XSConfidence Val:  0.9811 

----------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------
High Confidence value found:
From: sgithens@caret.cam.ac.uk

To: source@collab.sakaipro

High Confidence value found:
From: zqian@umich.edu

To: source@collab.sakaiproject.org
 
Date: Mon, 5 Nov 2007 14:57:58 -0500
 
XSConfidence Val:  0.9847 

----------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------
High Confidence value found:
From: aaronz@vt.edu

To: source@collab.sakaiproject.org
 
Date: Mon, 5 Nov 2007 13:05:11 -0500
 
XSConfidence Val:  0.9862 

----------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------
High Confidence value found:
From: ajpoland@iupui.edu

To: source@collab.sakaiproject.org
 
Date: Mon, 5 Nov 2007 11:01:03 -0500
 
XSConfidence Val:  0.9825 

----------------------------------------------------------------------------------------------
--------------------------------------------------------

----------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------
High Confidence value found:
From: aaronz@vt.edu

To: source@collab.sakaiproject.org
 
Date: Fri, 19 Oct 2007 13:53:33 -0400
 
XSConfidence Val:  0.9843 

----------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------
High Confidence value found:
From: gsilver@umich.edu

To: source@collab.sakaiproject.org
 
Date: Fri, 19 Oct 2007 13:33:54 -0400
 
XSConfidence Val:  0.9872 

----------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------
High Confidence value found:
From: gsilver@umich.edu

To: source@collab.sakaiproject.org
 
Date: Fri, 19 Oct 2007 12

#### **Exercise 5**

###### sections to reference: [list methods](https://www.py4e.com/html3/08-lists)

Below are the lines of a program I wrote to add some things to a list of my favorite foods, sort the list, and then print them out in alphabetical order. The problem is that, when I was uploading this program to the supercomputer, my internet stopped working and my program's lines got all mixed up.

Your task is this: rearrange the lines below so the program sorts and then prints the elements in my list of foods. When you've got the order all sorted out, put it in the code cell below. And thanks for your help!

```python
my_fave_foods.sort()
for food in my_fave_foods:
my_fave_foods.append("asiago cheese")
my_fave_foods = ["serrano peppers", "bananas", "apple pie", "veal", "smoked brisket"]
    print(food)
```

In [57]:
my_fave_foods = ["serrano peppers", "bananas", "apple pie", "veal", "smoked brisket"]
my_fave_foods.append("asiago cheese")
my_fave_foods.sort()
for food in my_fave_foods:
    print(food)

apple pie
asiago cheese
bananas
serrano peppers
smoked brisket
veal


#### **Exercise 6**

###### sections to reference: [lists](https://www.py4e.com/html3/08-lists)

So, now that you're an expert in sorting lists of food, let's get a little bit more literary. `richard3.txt` contains the text of Richard's opening soliloquy from Shakespeare's play *Richard III*. 

You can watch a performance of it [here](https://www.youtube.com/watch?v=OfaVYn1v4jM), if you're interested.

You're going to write a program that reads in the soliloquy from the file line by line. Then, using the `split` function, split each line into a list of words. Create a list to contain all of your words. Going word by word, check if the word is already in the list. If not, add it.

Once you've gone through the whole file, sort the list of words in alphabetical order, and print it out.

In [64]:
fn = "C:/Users/tatty/0010Github/Python-Part3/richard3.txt"
file = open(fn)
sameWord = 0
words = []
for line in file.readlines():
    for word in line.split(" "):
        if word in words:
            sameWord+=1
        else:
            words.append(word)
            
words.sort()
for word in words:
    print(word)
print("Duplicate word count:", sameWord)

About
And
As
But
By
Cheated
Clarence
Deformed
Dive
Edward
Edward's
G

Grim
Have
He
I
In
Into
King
Made
Nor
Now
Of
Our
Plots
That
This
To
Unless
Why
York

a
adversaries

against
alarums
all
am
ambling
amorous
an
and
are
arms
as
at
away
barded
bark
be
be

before
bosom
bound
breathing
brother
brows
bruised
buried

by
cannot
capers
chamber

changed
closely
clouds
comes
court
curtail'd
dangerous

day
days

deadly
deep
deformity

delight
delightful
descant
determined
discontent

dissembling
dogs
down
dreadful
dreams

drunken
entertain
fair
false
fearful
feature
for
fright
front

glass

glorious
half
halt
hate
hath
have
heirs
here

his
house

hung
idle
if
in
inductions
instead
is
just

king

lady's
laid
lamely
lascivious
libels
looking
lour'd
love's
lover

lute

made
majesty

marches
me
measures

meetings

merry
mew'd
mine
monuments

mounting
murderer
my
nature

nimbly
no
not
now
nymph

ocean
of
on
one
other

our
own
pass
peace

piping
pleasing
pleasures
prophecies
prophecy
proportion

prove


#### **Exercise 7**

###### sections to reference: [dictionaries and files](https://www.py4e.com/html3/09-dictionaries)

In the previous exercise, you just discarded words that were duplicates, but what if you wanted to count them?

You're going to read in `richard3.txt` again, but this time use a *dictionary*. Iterate through the lines in the file and, for each word, if this is its first occurence, add the word as a *key* to the dictionary and the number "1" as a *value*. If the word exists in the dictionary, *add one* to its value.

Then, print out your word-frequency dictionary.

In [70]:
fn = "C:/Users/tatty/0010Github/Python-Part3/richard3.txt"
file = open(fn)
sameWord = 0
words = dict()
#go over each line in the file
for line in file.readlines():
    #iterate over each word in the line
    for word in line.split(" "):
        #check if its already in the list
        if word in words:
            value = words[word]
            words[word] += value
        #if not add it to the list
        else:
            words[word] = 1
#print out the dict key value pairs
for word in words:
    print("Word:",word,":# used:", words[word])

Word: Now :# used: 2
Word: is :# used: 1
Word: the :# used: 4096
Word: winter :# used: 1
Word: of :# used: 512
Word: our :# used: 4
Word: discontent
 :# used: 1
Word: Made :# used: 1
Word: glorious :# used: 1
Word: summer :# used: 1
Word: by :# used: 4
Word: this :# used: 8
Word: sun :# used: 1
Word: York
 :# used: 1
Word: And :# used: 64
Word: all :# used: 1
Word: clouds :# used: 1
Word: that :# used: 32
Word: lour'd :# used: 1
Word: upon :# used: 1
Word: house
 :# used: 1
Word: In :# used: 2
Word: deep :# used: 1
Word: bosom :# used: 1
Word: ocean :# used: 1
Word: buried
 :# used: 1
Word: are :# used: 1
Word: brows :# used: 1
Word: bound :# used: 1
Word: with :# used: 1
Word: victorious :# used: 1
Word: wreaths
 :# used: 1
Word: Our :# used: 4
Word: bruised :# used: 1
Word: arms :# used: 1
Word: hung :# used: 1
Word: up :# used: 1
Word: for :# used: 2
Word: monuments
 :# used: 1
Word: stern :# used: 1
Word: alarums :# used: 1
Word: changed :# used: 1
Word: to :# used: 64
Word: merry 